R/ml_estimation.R

```diff
@@ -141,510 +161,211 @@ ml_estimation.default <- function(x,
-      estimates_loc_sc[[1]] + stats::qnorm((1 + conf_level) / 2) * se_loc_sc[[1]])
-
-    w <- exp(
-      (stats::qnorm((1 + conf_level) / 2) * se_loc_sc[[2]]) / estimates_loc_sc[[2]])
-    conf_sig <- c(estimates_loc_sc[[2]] /  w, estimates_loc_sc[[2]] * w)
-
-    conf_ints_loc_sc <- matrix(c(conf_mu, conf_sig), byrow = TRUE, ncol = 2)
-    colnames(conf_ints_loc_sc) <- c(paste(((1 - conf_level) / 2) * 100,"%"),
-                                    paste(((1 + conf_level) / 2) * 100, "%"))
-    rownames(conf_ints_loc_sc) <- names(estimates_loc_sc)
-
-    ml_output <- list(
-      coefficients = estimates_loc_sc,
-      confint = conf_ints_loc_sc,
-      varcov = vcov_loc_sc,
-      logL = ml$value,
-      aic = -2 * (ml$value) + 2 * length(estimates_loc_sc),
-      bic = (-2 * (ml$value) +
-               log(length(x)) * length(estimates_loc_sc))
+
+    ## Optimization of `loglik_profiling_()`:
+    opt_thres <- stats::optim(
+      par = t0,
+      fn = loglik_profiling_,
+      method = "L-BFGS-B",
+      lower = 0,
+      upper = (1 - (1 / 1e+5)) * min(x),
+      control = list(fnscale = -1), # no user input for profiling!
+      x = x,
+      status = status,
+      wts = wts,
+      distribution = distribution
     )
+
+    opt_thres <- opt_thres$par
+
+    ## Preparation for ml:
+    x <- x - opt_thres
   }
 
-  class(ml_output) <- c(
-    "wt_model", "wt_ml_estimation", "wt_model_estimation", class(ml_output)
+  # Step 1: Estimation of a two-parametric model (x or x - thres) using ML:
+
+  ## Force maximization:
+  control$fnscale <- -1
+
+  ## Use log scale:
+  start_dist_params[2] <- log(start_dist_params[2])
+
+  ## Two-parametric model:
+  ml <- stats::optim(
+    par = start_dist_params,
+    fn = loglik_function_,
+    method = "BFGS",
+    control = control,
+    hessian = TRUE,
+    x = x,
+    status = status,
+    wts = wts,
+    distribution = distribution,
+    log_scale = TRUE
   )
 
-  if (all(id == "", na.rm = TRUE)) {
-    ml_output$data <- tibble::tibble(
-      x = x, status = status
+  ## Parameters:
+  dist_params <- ml$par
+  names(dist_params) <- c("mu", "sigma")
+
+  ### Three-parametric, needed for variance-covariance matrix:
+  if (exists("opt_thres")) {
+    dist_params <- c(dist_params, opt_thres)
+
+    ml <- stats::optim(
+      par = dist_params,
+      fn = loglik_function_,
+      method = "BFGS",
+      control = control,
+      hessian = TRUE,
+      x = xx,
+      status = d,
+      wts = wts,
+      distribution = distribution,
+      log_scale = TRUE
     )
-  } else {
-    ml_output$data <- data
+
+    #### Update parameters:
+    dist_params <- ml$par
+    names(dist_params) <- c("mu", "sigma", "gamma")
   }
 
-  ml_output$distribution <- distribution
+  ## Value of the log-likelihood at optimum:
+  logL <- ml$value
 
-  return(ml_output)
-}
+  ## Variance-covariance matrix on log scale:
+  dist_varcov_logsigma <- solve(-ml$hessian)
```

Comment:
Commentary that this computes the inverse. This can be read in documentation of `solve`, but I didn't know this before